### PR TITLE
Changes from `smo-lunakit`

### PR DESCRIPTION
**Pull request by MonsterDruide1**
Adds all infos collected in the context of the mod `smo-lunakit`, by @Amethyst-szs, but probably also contains changes originally made by @CraftyBoss.

List of changes/"overrides" currently available at https://github.com/MonsterDruide1/smo-lunakit/tree/stable/lib/custom/nn, although this link will be invalid after this PR has been merged.

- - -
This change is [Reviewable](https://reviewable.io/reviews/open-ead/nnheaders/27)

### CODE REVIEW

File: include/nn/socket.h [hunk 1]
```diff
@@ -1,41 +1,177 @@
...
 #include <sys/socket.h>
 
-struct in_addr;
-
 namespace nn {
+class TimeSpan;
 namespace socket {
+struct ResourceStatistics;
 
-#if NN_SDK_VER >= NN_MAKE_VER(7, 0, 0)
```
Comment (leoetlino) [inline include/nn/socket.h:17]: backward compatibility?

File: include/nn/vfx/Resource.h [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    void BindExternalResTextureFile(gfx::ResTextureFile*);
+    void UnregisterTextureViewFromDescriptorPool(
+        void (*)(gfx::DescriptorSlot*, const gfx::TTextureView<gfx::ApiVariationNvn8>&, void*),
+        void*);
+    void BindExternalG3dResFile(g3d::ResFile*);
+    void BindExternalResShaderFile(gfx::ResShaderFile*, gfx::ResShaderFile*);
+    void InitializeEmitterSetResource(EmitterSetResource*, detail::BufferSizeCalculator*);
+    void InitializeEmitterResource(EmitterResource*, detail::BinaryData*, int,
+                                   detail::BufferSizeCalculator*);
+    void GetPrimitive(ulong) const;
```
Comment (leoetlino) [inline include/nn/vfx/Resource.h:53]: ulong should probably be size_t (or less likely, unsigned long / u64).

**Review by leoetlino — CHANGES_REQUESTED** [review]

File: include/nn/vfx/Resource.h [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+namespace detail {
+struct BinaryData;
+struct BufferSizeCalculator;
+struct ResNameTableRecord;
+}  // namespace detail
+
+class Resource {
+public:
+    Resource(gfx::TDevice<gfx::ApiVariationNvn8>*, Heap*, void*,
```
[the remark this answers is not in the document]
Comment (MonsterDruide1) [reply include/nn/vfx/Resource.h:25]: There is an [NVNdevice](https://code.botw.link/uking/uking/lib/NintendoSDK/include/nvn/nvn.h.html#NVNdevice), which is not the correct typedef though. It seems like so far, `TDevice<ApiVariationNvn8>` has been used manually (for example [here](https://code.botw.link/uking/uking/lib/NintendoSDK/include/nn/gfx/gfx_Interoperation-api.nvn.8.h.html))

**Comment by MonsterDruide1** [comment]
I just noticed that `swkbd.h` was already included in the repo, so I merged the new file into the existing one.

File: include/nn/vfx/Resource.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (ThePixelGamer) [reply include/nn/vfx/Resource.h:25]: https://github.com/open-ead/nnheaders/blob/master/include/nn/gfx/gfx_Types.h#L10

[the remark this answers is not in the document]
Comment (MonsterDruide1) [reply include/nn/vfx/Resource.h]: Ah, alright - adjusted accordingly, thanks for the reference!

**Review by leoetlino — COMMENTED** [review]
> Reviewed 5 of 6 files at r2.
> *[Reviewable](https://reviewable.io/reviews/open-ead/nnheaders/27)* status: 5 of 18 files reviewed, 3 unresolved discussions (waiting on @MonsterDruide1 and @ThePixelGamer)

___
*[`include/nn/swkbd/swkbd.h` line 7 at r2](https://reviewable.io/reviews/open-ead/nnheaders/27#-NtEEJJK1aepPfMJv6vl:-NtEEJJK1aepPfMJv6vm:b10n647) ([raw file](https://github.com/open-ead/nnheaders/blob/cb6449b86012e890cb8a1e097d72f8c34480b796/include/nn/swkbd/swkbd.h#L7)):*
> ```objc
> 
> namespace nn::applet {
> enum class ExitReason { Normal = 0, Canceled = 1, Abnormal = 2, Unexpected = 10 };
> ```

could you move this into an applet header?
___
*[`include/nn/swkbd/swkbd.h` line 167 at r2](https://reviewable.io/reviews/open-ead/nnheaders/27#-NtEEoa14qt6MBqXQLAy:-NtEEoa14qt6MBqXQLAz:b-iap3vw) ([raw file](https://github.com/open-ead/nnheaders/blob/cb6449b86012e890cb8a1e097d72f8c34480b796/include/nn/swkbd/swkbd.h#L167)):*
> ```objc
> ulong GetRequiredWorkBufferSize(bool);
> ulong GetRequiredStringBufferSize();
> nn::applet::ExitReason getExitReason();
> ```

GetExitReason (uppercase G)?

**Review by MonsterDruide1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/open-ead/nnheaders/27)* status: 4 of 19 files reviewed, 3 unresolved discussions (waiting on @leoetlino and @ThePixelGamer)

___
*[`include/nn/swkbd/swkbd.h` line 7 at r2](https://reviewable.io/reviews/open-ead/nnheaders/27#-NtEEJJK1aepPfMJv6vl:-Nv7qFHECkffZ6a8U-D7:b-896fix) ([raw file](https://github.com/open-ead/nnheaders/blob/cb6449b86012e890cb8a1e097d72f8c34480b796/include/nn/swkbd/swkbd.h#L7)):*
<details><summary><i>Previously, leoetlino (Léo Lam) wrote…</i></summary><blockquote>

could you move this into an applet header?
</blockquote></details>

Done.
___
*[`include/nn/swkbd/swkbd.h` line 167 at r2](https://reviewable.io/reviews/open-ead/nnheaders/27#-NtEEoa14qt6MBqXQLAy:-Nv7qFnE63OFwJCQzL2k:b-896fix) ([raw file](https://github.com/open-ead/nnheaders/blob/cb6449b86012e890cb8a1e097d72f8c34480b796/include/nn/swkbd/swkbd.h#L167)):*
<details><summary><i>Previously, leoetlino (Léo Lam) wrote…</i></summary><blockquote>

GetExitReason (uppercase G)?
</blockquote></details>

Done.
___
*[`include/nn/vfx/Resource.h` line 53 at r1](https://reviewable.io/reviews/open-ead/nnheaders/27#gh-1523960376:-Nv7qHujEGcdE-ePm1AK:b-896fix) ([raw file](https://github.com/open-ead/nnheaders/blob/6aad273f830c314bf721f0258ca1a9dca9139057/include/nn/vfx/Resource.h#L53)):*
<details><summary><i>Previously, leoetlino (Léo Lam) wrote…</i></summary><blockquote>

ulong should probably be size_t (or less likely, unsigned long / u64).
</blockquote></details>

Done.

**Review by leoetlino — APPROVED** [review]
> Reviewed 12 of 16 files at r1, 1 of 6 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/open-ead/nnheaders/27)* status: all files reviewed, 2 unresolved discussions (waiting on @MonsterDruide1 and @ThePixelGamer)

___
*[`include/nn/diag.h` line 28 at r3](https://reviewable.io/reviews/open-ead/nnheaders/27#-Nw2AWqnCe5XPLpS33EX:-Nw2AWqnCe5XPLpS33EY:b-7td3tf) ([raw file](https://github.com/open-ead/nnheaders/blob/f2846888854e55e964a2ec4834912cda75dd8c98/include/nn/diag.h#L28)):*
> ```objc
> 
> // MODULE / SYMBOL
> u64 GetSymbolName(char* name, u64 nameSize, u64 addr);
> ```

I think this should probably be taking a size_t nameSize and returning a size_t instead of u64. addr should probably be uintptr_t
___
*[`include/nn/fs/fs_directories.h` line 34 at r3](https://reviewable.io/reviews/open-ead/nnheaders/27#-Nw2Axf83bXUcpPHnH71:-Nw2Axf83bXUcpPHnH72:b-e0y07i) ([raw file](https://github.com/open-ead/nnheaders/blob/f2846888854e55e964a2ec4834912cda75dd8c98/include/nn/fs/fs_directories.h#L34)):*
> ```objc
>     path: Path to path to create.
> */
> Result CreateDirectory(char const* path);
> ```

const char*

**Review by MonsterDruide1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/open-ead/nnheaders/27)* status: all files reviewed, 2 unresolved discussions (waiting on @leoetlino and @ThePixelGamer)

___
*[`include/nn/diag.h` line 28 at r3](https://reviewable.io/reviews/open-ead/nnheaders/27#-Nw2AWqnCe5XPLpS33EX:-NwLp-0C5kgbhnDAzGRe:bdr74g7) ([raw file](https://github.com/open-ead/nnheaders/blob/f2846888854e55e964a2ec4834912cda75dd8c98/include/nn/diag.h#L28)):*
<details><summary><i>Previously, leoetlino (Léo Lam) wrote…</i></summary><blockquote>

I think this should probably be taking a size_t nameSize and returning a size_t instead of u64. addr should probably be uintptr_t
</blockquote></details>

I don't know the difference between those types, so you might be right - I just worked based on the symbols, not sure if the types you suggested "compile back" to the types I annotated here. Changed it though.
___
*[`include/nn/fs/fs_directories.h` line 34 at r3](https://reviewable.io/reviews/open-ead/nnheaders/27#-Nw2Axf83bXUcpPHnH71:-NwLpVDUE2AKi68DI-gE:bruiwv6) ([raw file](https://github.com/open-ead/nnheaders/blob/f2846888854e55e964a2ec4834912cda75dd8c98/include/nn/fs/fs_directories.h#L34)):*
<details><summary><i>Previously, leoetlino (Léo Lam) wrote…</i></summary><blockquote>

const char*
</blockquote></details>

Adjusted here and in a few other files I encountered.

**Review by leoetlino — APPROVED** [review]
> Reviewed 4 of 4 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/open-ead/nnheaders/27)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @ThePixelGamer)

**Review by ThePixelGamer — APPROVED** [review]
LGTM

**Pull request merged**